File: core/src/main/java/org/toml/lang/psi/TomlVisitor.java
/*
 * Use of this source code is governed by the MIT license that can be
 * found in the LICENSE file.
 */

package org.toml.lang.psi;

import consulo.language.psi.PsiElementVisitor;

public class TomlVisitor extends PsiElementVisitor {
    public void visitElement(TomlElement element) {
        super.visitElement(element);
    }

    public void visitValue(TomlValue element) {
        visitElement(element);
    }

    public void visitKeyValue(TomlKeyValue element) {
        visitElement(element);
    }

    public void visitKeySegment(TomlKeySegment element) {
        visitElement(element);
    }

    public void visitKey(TomlKey element) {
        visitElement(element);
    }

    public void visitLiteral(TomlLiteral element) {
        visitValue(element);
    }

    public void visitKeyValueOwner(TomlKeyValueOwner element) {
        visitElement(element);
    }

    public void visitArray(TomlArray element) {
        visitValue(element);
    }

    public void visitTable(TomlTable element) {
        visitKeyValueOwner(element);
    }

    public void visitTableHeader(TomlTableHeader element) {
        visitElement(element);
    }

    public void visitInlineTable(TomlInlineTable element) {
        visitKeyValueOwner(element);
    }

    public void visitArrayTable(TomlArrayTable element) {
        visitKeyValueOwner(element);
    }
}

File: core/src/main/kotlin/org/toml/lang/psi/TomlVisitor.kt
/*
 * Use of this source code is governed by the MIT license that can be
 * found in the LICENSE file.
 */

package org.toml.lang.psi

import consulo.application.progress.ProgressManager
import consulo.language.psi.PsiElement
import consulo.language.psi.PsiElementVisitor
import consulo.language.psi.PsiRecursiveVisitor

open class TomlVisitor : PsiElementVisitor() {
    open fun visitElement(element: TomlElement) {
        super.visitElement(element)
    }

    open fun visitValue(element: TomlValue) {
        visitElement(element)
    }

    open fun visitKeyValue(element: TomlKeyValue) {
        visitElement(element)
    }

    open fun visitKeySegment(element: TomlKeySegment) {
        visitElement(element)
    }

    open fun visitKey(element: TomlKey) {
        visitElement(element)
    }

    open fun visitLiteral(element: TomlLiteral) {
        visitValue(element)
    }

    open fun visitKeyValueOwner(element: TomlKeyValueOwner) {
        visitElement(element)
    }

    open fun visitArray(element: TomlArray) {
        visitValue(element)
    }

    open fun visitTable(element: TomlTable) {
        visitKeyValueOwner(element)
    }

    open fun visitTableHeader(element: TomlTableHeader) {
        visitElement(element)
    }

    open fun visitInlineTable(element: TomlInlineTable) {
        visitKeyValueOwner(element)
    }

    open fun visitArrayTable(element: TomlArrayTable) {
        visitKeyValueOwner(element)
    }
}

open class TomlRecursiveVisitor : TomlVisitor(), PsiRecursiveVisitor {
    override fun visitElement(element: PsiElement) {
        ProgressManager.checkCanceled()
        element.acceptChildren(this)
    }

    override fun visitElement(element: TomlElement) {
        visitElement(element as PsiElement)
    }
}
